core/templates/dev/head/pages/community-dashboard-page/services/translate-text.service.spec.ts

```diff
@@ -0,0 +1,86 @@
+// Copyright 2018 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Tests that the translate-text service is working as
+ * expected.
+ */
+
+require(
+  'pages/community-dashboard-page/services/' +
+  'translate-text.service.ts');
+
+describe('TranslateTextService', function() {
+  var TranslateTextService, $httpBackend;
+  var mockTranslateBackendObject;
+  var callbackObject;
+  beforeEach(angular.mock.module('oppia'));
+  beforeEach(angular.mock.inject(function($injector) {
+    $httpBackend = $injector.get('$httpBackend');
+    TranslateTextService = $injector.get('TranslateTextService');
+    mockTranslateBackendObject = {
+      version: 1,
+      state_names_to_content_id_mapping: {
+        Intro: {
+          content: 'IntroContent'
+        },
+        END: {
+          content: '<p>Congratulations, you have finished!</p>'
+        },
+        'Second Question': {
+          content: 'testContent'
+        }
+      }
+    };
+  }));
+
+  afterEach(function() {
+    $httpBackend.verifyNoOutstandingExpectation();
+    $httpBackend.verifyNoOutstandingRequest();
+  });
+
+  describe('getTextToTranslate', function() {
+    it('should return text available for translation', function() {
+      $httpBackend.whenRoute('GET', '/gettranslatabletexthandler')
+        .respond(function(method, url, data, headers, params) {
+          return [200, mockTranslateBackendObject];
+        });
+      TranslateTextService.init(1, 'en', function() {
+        var translateData = TranslateTextService.getTextToTranslate();
+        expect(translateData.text).toEqual('IntroContent');
+        expect(translateData.more).toEqual(true);
+      });
+      $httpBackend.flush();
+    });
+  });
+
+  describe('suggestTranslatedText', function() {
+    it('should post new translated text data to the API', function() {
+      callbackObject = {
+        callback: function(data) {}
+      };
+      spyOn(callbackObject, 'callback');
```

Comment:
create two spyOn, one for success and another for failure and check success is called and failure is "not called".

Reply:
But the original function (which wasn't written by me) does not have a failure callback. Should I modify it?

Reply:
@DubeySandeep also, using spyOn is problematic here, as the callback function here is passed through multiple functions as callbacks - so even though it's called, its not detected in toHaveBeenCalled. Getting around this would require me to change the source code (and I'm not sure if it can be done even then). 

Should I just test these functions by verifying some data in the service after their execution?

Reply:
Or if there is a way to spy callbacks, can you point me to it?

Reply:
maybe the spyOn test would be for the *backend-api-service.spec file and the translate-test.service can be independent of it? (not sure!)

Reply:
I don't see how that would work out. The backend service only makes API calls and passes on data. There's no internal processing to test.